File: app/src/main/java/com/ash/studios/musify/Services/MusicService.java
package com.ash.studios.musify.Services;

import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaMetadata;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.IBinder;
import android.provider.MediaStore;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.ash.studios.musify.Activities.Player;
import com.ash.studios.musify.Interfaces.IControl;
import com.ash.studios.musify.Interfaces.IService;
import com.ash.studios.musify.Models.Song;
import com.ash.studios.musify.R;
import com.ash.studios.musify.utils.App;
import com.ash.studios.musify.utils.Constants;
import com.ash.studios.musify.utils.Engine;
import com.ash.studios.musify.utils.Instance;
import com.ash.studios.musify.utils.Utils;

import static com.ash.studios.musify.utils.App.CHANNEL_ID;

public class MusicService extends Service implements MediaPlayer.OnCompletionListener {
    Context context;
    MediaSessionCompat mediaSession;

    @Override
    public void onCreate() {
        super.onCreate();
        context = this;
        mediaSession = new MediaSessionCompat(context, "MUSIFY");
        mediaSession.setMetadata(MediaMetadataCompat.fromMediaMetadata(
                new MediaMetadata.Builder().putLong(MediaMetadata.METADATA_KEY_DURATION, -1L).build()
        ));
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        switch (intent.getAction()) {
            case Constants.ACTION.CREATE:
                if (Instance.playing) notificationForPlay();
                else notificationForPause();
                break;
            case Constants.ACTION.STOP_SERVICE:
                ((IService) ((App) getApplicationContext()).getCurrentContext()).onStopService();
                stopSelf();
                break;
            case Constants.ACTION.PREV:
                ((IControl) ((App) getApplicationContext()).getCurrentContext()).onPrevClicked();
                break;
            case Constants.ACTION.PLAY:
                Instance.playing = true;
                ((IControl) ((App) getApplicationContext()).getCurrentContext()).onPlayClicked();
                notificationForPlay();
                break;
            case Constants.ACTION.PAUSE:
                Instance.playing = false;
                ((IControl) ((App) getApplicationContext()).getCurrentContext()).onPauseClicked();
                notificationForPause();
                break;
            case Constants.ACTION.NEXT:
                ((IControl) ((App) getApplicationContext()).getCurrentContext()).onNextClicked();
                break;
        }

        return START_NOT_STICKY;
    }

    private void notificationForPlay() {
        Intent mainIntent = new Intent(context, Player.class)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent mainPending = PendingIntent.getActivity(context, 0, mainIntent, 0);

        Intent stopIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.STOP_SERVICE);
        PendingIntent stopPending = PendingIntent.getService(context, 0, stopIntent, 0);

        Intent prevIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.PREV);
        PendingIntent prevPending = PendingIntent.getService(context, 0, prevIntent, 0);

        Intent pauseIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.PAUSE);
        PendingIntent pausePending = PendingIntent.getService(context, 0, pauseIntent, 0);

        Intent nextIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.NEXT);
        PendingIntent nextPending = PendingIntent.getService(context, 0, nextIntent, 0);

        Song song = Instance.songs.get(Instance.position);
        StringBuilder count = new StringBuilder().append(Instance.position + 1).append("/").append(Instance.songs.size());
        Notification notification = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)

                .addAction(R.drawable.ic_previous, "Previous", prevPending)
                .addAction(R.drawable.ic_pause, "Pause", pausePending)
                .addAction(R.drawable.ic_next, "Next", nextPending)
                .addAction(R.drawable.ic_close, "Stop", stopPending)

                .setStyle(new androidx.media.app.NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(1, 2, 3)
                        .setMediaSession(mediaSession.getSessionToken()))
                .setLargeIcon(getBitmap(Utils.getAlbumArt(song.getAlbum_id())))
                .setSmallIcon(R.drawable.ic_icon)

                .setContentTitle(song.getTitle())
                .setContentText(song.getArtist())
                .setSubText(count)

                .setContentIntent(mainPending)
                .build();

        startForeground(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE, notification);
    }

    private void notificationForPause() {
        Intent mainIntent = new Intent(context, Player.class)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        PendingIntent mainPending = PendingIntent.getActivity(context, 0, mainIntent, 0);

        Intent stopIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.STOP_SERVICE);
        PendingIntent stopPending = PendingIntent.getService(context, 3, stopIntent, 0);

        Intent prevIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.PREV);
        PendingIntent prevPending = PendingIntent.getService(context, 0, prevIntent, 0);

        Intent playIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.PLAY);
        PendingIntent playPending = PendingIntent.getService(context, 1, playIntent, 0);

        Intent nextIntent = new Intent(context, MusicService.class).setAction(Constants.ACTION.NEXT);
        PendingIntent nextPending = PendingIntent.getService(context, 2, nextIntent, 0);

        Song song = Instance.songs.get(Instance.position);
        StringBuilder count = new StringBuilder().append(Instance.position + 1).append("/").append(Instance.songs.size());
        Notification notification = new NotificationCompat.Builder(context, CHANNEL_ID)
                .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)

                .addAction(R.drawable.ic_previous, "Previous", prevPending)
                .addAction(R.drawable.ic_play, "Play", playPending)
                .addAction(R.drawable.ic_next, "Next", nextPending)
                .addAction(R.drawable.ic_close, "Stop", stopPending)

                .setStyle(new androidx.media.app.NotificationCompat.MediaStyle()
                        .setShowActionsInCompactView(1, 2, 3)
                        .setMediaSession(mediaSession.getSessionToken()))
                .setLargeIcon(getBitmap(Utils.getAlbumArt(song.getAlbum_id())))
                .setSmallIcon(R.drawable.ic_icon)

                .setContentTitle(song.getTitle())
                .setContentText(song.getArtist())
                .setSubText(count)

                .setContentIntent(mainPending)
                .build();

        startForeground(Constants.NOTIFICATION_ID.FOREGROUND_SERVICE, notification);
    }

    private Bitmap getBitmap(Uri uri) {
        Bitmap bitmap;
        try {
            bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), uri);
        } catch (Exception e) {
            bitmap = BitmapFactory.decodeResource(getResources(), R.drawable.placeholder);
        }
        return bitmap;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCompletion(MediaPlayer mediaPlayer) {
        new Engine(context).playNextSong();
        if (Instance.mp != null) {
            Instance.mp = MediaPlayer.create(getApplicationContext(), Instance.uri);
            Instance.mp.start();
            Instance.mp.setOnCompletionListener(this);
        }
        Utils.putCurrentPosition(context, Instance.position);
        notificationForPlay();
    }
}

File: app/src/main/java/com/ash/studios/musify/services/MusicService.kt
package com.ash.studios.musify.services

import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadata
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.net.Uri
import android.os.IBinder
import android.provider.MediaStore
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import androidx.media.app.NotificationCompat
import com.ash.studios.musify.R
import com.ash.studios.musify.activities.Player
import com.ash.studios.musify.interfaces.IControl
import com.ash.studios.musify.interfaces.IService
import com.ash.studios.musify.models.Song
import com.ash.studios.musify.utils.App
import com.ash.studios.musify.utils.Constants
import com.ash.studios.musify.utils.Constants.FOREGROUND_SERVICE
import com.ash.studios.musify.utils.Engine
import com.ash.studios.musify.utils.Instance.mp
import com.ash.studios.musify.utils.Instance.playing
import com.ash.studios.musify.utils.Instance.position
import com.ash.studios.musify.utils.Instance.songs
import com.ash.studios.musify.utils.Instance.uri
import com.ash.studios.musify.utils.Utils

class MusicService : Service(), OnCompletionListener {
    private lateinit var context: Context
    private lateinit var mediaSession: MediaSessionCompat
    override fun onCreate() {
        super.onCreate()
        context = this
        mediaSession = MediaSessionCompat(context, "MUSIFY")
        mediaSession.setMetadata(
            MediaMetadataCompat.fromMediaMetadata(
                MediaMetadata.Builder().putLong(MediaMetadata.METADATA_KEY_DURATION, -1L).build()
            )
        )
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (Constants.ACTION.CREATE.label == intent.action) {
            if (playing) notificationForPlay() else notificationForPause()
        } else if (Constants.ACTION.STOP_SERVICE.label == intent.action) {
            ((applicationContext as App).currentContext as IService?)!!.onStopService()
            stopSelf()
        } else if (Constants.ACTION.PREV.label == intent.action) {
            ((applicationContext as App).currentContext as IControl?)!!.onPrevClicked()
        } else if (Constants.ACTION.PLAY.label == intent.action) {
            playing = true
            ((applicationContext as App).currentContext as IControl?)!!.onPlayClicked()
            notificationForPlay()
        } else if (Constants.ACTION.PAUSE.label == intent.action) {
            playing = false
            ((applicationContext as App).currentContext as IControl?)!!.onPauseClicked()
            notificationForPause()
        } else if (Constants.ACTION.NEXT.label == intent.action) {
            ((applicationContext as App).currentContext as IControl?)!!.onNextClicked()
        }
        return START_NOT_STICKY
    }

    private fun notificationForPlay() {
        val mainIntent = Intent(context, Player::class.java)
            .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        val mainPending = PendingIntent.getActivity(context, 0, mainIntent, 0)
        val stopIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.STOP_SERVICE.label)
        val stopPending = PendingIntent.getService(context, 0, stopIntent, 0)
        val prevIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.PREV.label)
        val prevPending = PendingIntent.getService(context, 0, prevIntent, 0)
        val pauseIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.PAUSE.label)
        val pausePending = PendingIntent.getService(context, 0, pauseIntent, 0)
        val nextIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.NEXT.label)
        val nextPending = PendingIntent.getService(context, 0, nextIntent, 0)
        val song: Song = songs!![position]
        val count: StringBuilder = StringBuilder().append(position + 1).append("/").append(songs!!.size)
        val notification = androidx.core.app.NotificationCompat.Builder(context, App.CHANNEL_ID)
            .setVisibility(androidx.core.app.NotificationCompat.VISIBILITY_PUBLIC)
            .addAction(R.drawable.ic_previous, "Previous", prevPending)
            .addAction(R.drawable.ic_pause, "Pause", pausePending)
            .addAction(R.drawable.ic_next, "Next", nextPending)
            .addAction(R.drawable.ic_close, "Stop", stopPending)
            .setStyle(
                NotificationCompat.MediaStyle()
                    .setShowActionsInCompactView(1, 2, 3)
                    .setMediaSession(mediaSession.sessionToken)
            )
            .setLargeIcon(getBitmap(Utils.getAlbumArt(song.album_id)))
            .setSmallIcon(R.drawable.ic_icon)
            .setContentTitle(song.title)
            .setContentText(song.artist)
            .setSubText(count)
            .setContentIntent(mainPending)
            .build()
        startForeground(FOREGROUND_SERVICE, notification)
    }

    private fun notificationForPause() {
        val mainIntent = Intent(context, Player::class.java)
            .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
        val mainPending = PendingIntent.getActivity(context, 0, mainIntent, 0)
        val stopIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.STOP_SERVICE.label)
        val stopPending = PendingIntent.getService(context, 3, stopIntent, 0)
        val prevIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.PREV.label)
        val prevPending = PendingIntent.getService(context, 0, prevIntent, 0)
        val playIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.PLAY.label)
        val playPending = PendingIntent.getService(context, 1, playIntent, 0)
        val nextIntent = Intent(context, MusicService::class.java).setAction(Constants.ACTION.NEXT.label)
        val nextPending = PendingIntent.getService(context, 2, nextIntent, 0)
        val song: Song = songs!!.get(position)
        val count: StringBuilder = StringBuilder().append(position + 1).append("/").append(songs!!.size)
        val notification = androidx.core.app.NotificationCompat.Builder(context, App.CHANNEL_ID)
            .setVisibility(androidx.core.app.NotificationCompat.VISIBILITY_PUBLIC)
            .addAction(R.drawable.ic_previous, "Previous", prevPending)
            .addAction(R.drawable.ic_play, "Play", playPending)
            .addAction(R.drawable.ic_next, "Next", nextPending)
            .addAction(R.drawable.ic_close, "Stop", stopPending)
            .setStyle(
                NotificationCompat.MediaStyle()
                    .setShowActionsInCompactView(1, 2, 3)
                    .setMediaSession(mediaSession.sessionToken)
            )
            .setLargeIcon(getBitmap(Utils.getAlbumArt(song.album_id)))
            .setSmallIcon(R.drawable.ic_icon)
            .setContentTitle(song.title)
            .setContentText(song.artist)
            .setSubText(count)
            .setContentIntent(mainPending)
            .build()
        startForeground(FOREGROUND_SERVICE, notification)
    }

    private fun getBitmap(uri: Uri): Bitmap {
        val bitmap: Bitmap = try {
            MediaStore.Images.Media.getBitmap(contentResolver, uri)
        } catch (e: Exception) {
            BitmapFactory.decodeResource(resources, R.drawable.placeholder)
        }
        return bitmap
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCompletion(mediaPlayer: MediaPlayer) {
        Engine(context).playNextSong()
        if (mp != null) {
            mp = MediaPlayer.create(applicationContext, uri)
            mp!!.start()
            mp!!.setOnCompletionListener(this)
        }
        Utils.putCurrentPosition(context, position)
        notificationForPlay()
    }
}